In [30]:
import numpy as np
import pandas as pd


In [31]:
df1=pd.read_csv("website_classification.csv")
df1.head()

Unnamed: 0                                        website_url  \
0           0     https://www.booking.com/index.html?aid=1743217   
1           1                   https://travelsites.com/expedia/   
2           2               https://travelsites.com/tripadvisor/   
3           3              https://www.momondo.in/?ispredir=true   
4           4  https://www.ebookers.com/?AFFCID=EBOOKERS-UK.n...   

                                cleaned_website_text Category  
0  official site good hotel accommodation big sav...   Travel  
1  expedia hotel book sites like use vacation wor...   Travel  
2  tripadvisor hotel book sites like previously d...   Travel  
3  cheap flights search compare flights momondo f...   Travel  
4  bot create free account create free account si...   Travel

In [32]:
df2=pd.read_csv("youtube_thumbnail.csv")
df2.head()

Id      Channel Category  \
0  OkmNXy7er84  3Blue1Brown  Science   
1  r6sGWTCMz2k  3Blue1Brown  Science   
2  bBC-nXj3Ng4  3Blue1Brown  Science   
3  aircAruvnKk  3Blue1Brown  Science   
4  HEfHFsfGXjs  3Blue1Brown  Science   

                                               Title  
0            The hardest problem on the hardest test  
1  But what is a Fourier series? From heat flow t...  
2                But how does bitcoin actually work?  
3  But what is a neural network? | Chapter 1, Dee...  
4    The most unexpected answer to a counting puzzle

In [33]:
df1.head()

Unnamed: 0                                        website_url  \
0           0     https://www.booking.com/index.html?aid=1743217   
1           1                   https://travelsites.com/expedia/   
2           2               https://travelsites.com/tripadvisor/   
3           3              https://www.momondo.in/?ispredir=true   
4           4  https://www.ebookers.com/?AFFCID=EBOOKERS-UK.n...   

                                cleaned_website_text Category  
0  official site good hotel accommodation big sav...   Travel  
1  expedia hotel book sites like use vacation wor...   Travel  
2  tripadvisor hotel book sites like previously d...   Travel  
3  cheap flights search compare flights momondo f...   Travel  
4  bot create free account create free account si...   Travel

In [34]:
df2.columns

Index(['Id', 'Channel', 'Category', 'Title'], dtype='object')

In [35]:
df1.columns

Index(['Unnamed: 0', 'website_url', 'cleaned_website_text', 'Category'], dtype='object')

In [36]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1408 entries, 0 to 1407
Data columns (total 4 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Unnamed: 0            1408 non-null   int64 
 1   website_url           1408 non-null   object
 2   cleaned_website_text  1408 non-null   object
 3   Category              1408 non-null   object
dtypes: int64(1), object(3)
memory usage: 44.1+ KB


In [37]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2515 entries, 0 to 2514
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Id        2515 non-null   object
 1   Channel   2515 non-null   object
 2   Category  2515 non-null   object
 3   Title     2515 non-null   object
dtypes: object(4)
memory usage: 78.7+ KB


In [38]:
df1['Category'].unique()

array(['Travel', 'Social Networking and Messaging', 'News',
       'Streaming Services', 'Sports', 'Photography',
       'Law and Government', 'Health and Fitness', 'Games', 'E-Commerce',
       'Forums', 'Food', 'Education', 'Computers and Technology',
       'Business/Corporate', 'Adult'], dtype=object)

In [39]:
df2['Category'].unique()

array(['Science', 'News', 'Food', 'Blog', 'Tech', 'Informative',
       'Comedy,Entertainment', 'Entertainment', 'Automobile',
       'Tech,Informative', 'Automobile,Comedy', 'VideoGames',
       'Food,Entertainment', 'Blog,Comedy', 'Comedy,Informative',
       'Tech,Comedy', 'Comedy', 'Blog,Science', 'Blog,Entertainment',
       'Entertainment,Comedy', 'Tech,News', 'Entertainment,Blog'],
      dtype=object)

In [40]:
df1['Category'].value_counts()

Category
Education                          114
Business/Corporate                 109
Travel                             107
Streaming Services                 105
Sports                             104
E-Commerce                         102
Games                               98
News                                96
Health and Fitness                  96
Computers and Technology            93
Photography                         93
Food                                92
Law and Government                  84
Social Networking and Messaging     83
Forums                              16
Adult                               16
Name: count, dtype: int64

In [41]:
df2['Category'].value_counts()

Category
Science                 323
Food                    281
VideoGames              265
News                    218
Informative             218
Automobile              197
Tech                    193
Blog                    160
Comedy,Entertainment     88
Entertainment,Comedy     88
Automobile,Comedy        85
Blog,Comedy              56
Comedy,Informative       56
Entertainment            55
Entertainment,Blog       30
Comedy                   30
Blog,Entertainment       29
Tech,Informative         29
Tech,News                29
Tech,Comedy              29
Food,Entertainment       28
Blog,Science             28
Name: count, dtype: int64

In [42]:
%pip install emoji nltk

In [43]:
import emoji
import re
import nltk
from nltk import PorterStemmer
from nltk.corpus import stopwords

nltk.download('stopwords')
nltk.download('punkt')

stop_words = set(stopwords.words('english'))

port_stemmer = PorterStemmer()
def preprocessing(text):
    text = text.lower()
    text = re.sub(r"http\S+", "", text)
    text = re.sub(r"[^a-zA-Z]", " ", text)
    text = re.sub(r"\s+", " ", text)
    text = re.sub(r'[^\w\s,.!?]', '', text)  # removes most emojis & symbols
    text = text.split()
    text = [port_stemmer.stem(word) for word in text if not word in stop_words]
    return text

df1["preprocessed_text"] = df1["cleaned_website_text"].apply(preprocessing)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [44]:
df2.head()

Id      Channel Category  \
0  OkmNXy7er84  3Blue1Brown  Science   
1  r6sGWTCMz2k  3Blue1Brown  Science   
2  bBC-nXj3Ng4  3Blue1Brown  Science   
3  aircAruvnKk  3Blue1Brown  Science   
4  HEfHFsfGXjs  3Blue1Brown  Science   

                                               Title  
0            The hardest problem on the hardest test  
1  But what is a Fourier series? From heat flow t...  
2                But how does bitcoin actually work?  
3  But what is a neural network? | Chapter 1, Dee...  
4    The most unexpected answer to a counting puzzle

In [45]:
df_out = df1.copy()
df_out= df_out[~df_out["Category"].isin(["Adult", "Forums"])]
df_out["Category"].value_counts()

Category
Education                          114
Business/Corporate                 109
Travel                             107
Streaming Services                 105
Sports                             104
E-Commerce                         102
Games                               98
Health and Fitness                  96
News                                96
Photography                         93
Computers and Technology            93
Food                                92
Law and Government                  84
Social Networking and Messaging     83
Name: count, dtype: int64

# Task
Perform text classification on the combined data from `df1` and `df2` to predict the 'Category' based on the 'Title' column. Preprocess the 'Title' column in `df2`, combine the dataframes, vectorize the text data using `TfidfVectorizer`, split the data into training and testing sets, train a Logistic Regression model, and evaluate the model's performance.

## Preprocess df2

### Subtask:
Apply the `preprocessing` function to the 'Title' column of `df2` and create a new column 'preprocessed_text'.


**Reasoning**:
Apply the preprocessing function to the 'Title' column of df2 and store the result in a new column named 'preprocessed_text'.



In [46]:
df2["preprocessed_text"] = df2["Title"].apply(preprocessing)
df2.head()

Id      Channel Category  \
0  OkmNXy7er84  3Blue1Brown  Science   
1  r6sGWTCMz2k  3Blue1Brown  Science   
2  bBC-nXj3Ng4  3Blue1Brown  Science   
3  aircAruvnKk  3Blue1Brown  Science   
4  HEfHFsfGXjs  3Blue1Brown  Science   

                                               Title  \
0            The hardest problem on the hardest test   
1  But what is a Fourier series? From heat flow t...   
2                But how does bitcoin actually work?   
3  But what is a neural network? | Chapter 1, Dee...   
4    The most unexpected answer to a counting puzzle   

                              preprocessed_text  
0             [hardest, problem, hardest, test]  
1  [fourier, seri, heat, flow, draw, circl, de]  
2                       [bitcoin, actual, work]  
3       [neural, network, chapter, deep, learn]  
4              [unexpect, answer, count, puzzl]

## Combine dataframes

### Subtask:
Vertically concatenate `df1` and `df2` into a single dataframe.


**Reasoning**:
Vertically concatenate df1 and df2 into a single dataframe and display the head to verify the concatenation.



In [47]:
df_combined = pd.concat([df1, df2], ignore_index=True)
display(df_combined.head())

Unnamed: 0                                        website_url  \
0         0.0     https://www.booking.com/index.html?aid=1743217   
1         1.0                   https://travelsites.com/expedia/   
2         2.0               https://travelsites.com/tripadvisor/   
3         3.0              https://www.momondo.in/?ispredir=true   
4         4.0  https://www.ebookers.com/?AFFCID=EBOOKERS-UK.n...   

                                cleaned_website_text Category  \
0  official site good hotel accommodation big sav...   Travel   
1  expedia hotel book sites like use vacation wor...   Travel   
2  tripadvisor hotel book sites like previously d...   Travel   
3  cheap flights search compare flights momondo f...   Travel   
4  bot create free account create free account si...   Travel   

                                   preprocessed_text   Id Channel Title  
0  [offici, site, good, hotel, accommod, big, sav...  NaN     NaN   NaN  
1  [expedia, hotel, book, site, like, use, vacat,...  NaN     NaN   NaN  
2  [tripadvisor, hotel, book, site, like, previou...  NaN     NaN   NaN  
3  [cheap, flight, search, compar, flight, momond...  NaN     NaN   NaN  
4  [bot, creat, free, account, creat, free, accou...  NaN     NaN   NaN

## Vectorize combined data

### Subtask:
Use `TfidfVectorizer` to convert the 'preprocessed_text' column of the combined dataframe into numerical features.


**Reasoning**:
Convert the list of preprocessed words into a string and then vectorize the text data using TfidfVectorizer.



In [48]:
from sklearn.feature_extraction.text import TfidfVectorizer

df_combined['preprocessed_text_str'] = df_combined['preprocessed_text'].apply(lambda x: ' '.join(x) if isinstance(x, list) else '')

tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(df_combined['preprocessed_text_str'])

print("TF-IDF matrix shape:", tfidf_matrix.shape)

TF-IDF matrix shape: (3923, 47615)


## Split data

### Subtask:
Split the vectorized data and the corresponding categories into training and testing sets.


**Reasoning**:
Split the vectorized data and categories into training and testing sets using train_test_split.



In [49]:
from sklearn.model_selection import train_test_split

X = tfidf_matrix
y = df_combined['Category']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (3138, 47615)
Shape of X_test: (785, 47615)
Shape of y_train: (3138,)
Shape of y_test: (785,)


## Train and evaluate model

### Subtask:
Train a Logistic Regression model on the training data and evaluate its performance using a classification report and accuracy score.


**Reasoning**:
Train a Logistic Regression model and evaluate its performance using a classification report and accuracy score.



In [50]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score

model = LogisticRegression()
model.fit(X_train, y_train)
predictions = model.predict(X_test)

print("Classification Report:")
print(classification_report(y_test, predictions))

print("Accuracy Score:")
print(accuracy_score(y_test, predictions))

Classification Report:
                                 precision    recall  f1-score   support

                          Adult       0.00      0.00      0.00         4
                     Automobile       0.98      0.88      0.92        48
              Automobile,Comedy       0.89      0.89      0.89         9
                           Blog       0.93      0.76      0.83        33
                    Blog,Comedy       1.00      0.08      0.14        13
             Blog,Entertainment       0.00      0.00      0.00         3
                   Blog,Science       1.00      0.75      0.86         8
             Business/Corporate       1.00      0.75      0.86        16
                         Comedy       1.00      1.00      1.00         8
           Comedy,Entertainment       0.95      1.00      0.98        20
             Comedy,Informative       1.00      0.25      0.40        12
       Computers and Technology       0.67      0.62      0.64        13
                     E-Comm

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [51]:
# Define the mapping for the new categories
category_mapping = {
    'Education': 'study',
    'Science': 'study',
    'Computers and Technology': 'study',
    'Tech': 'study',
    'Informative': 'study',
    'Blog,Science': 'study',
    'Tech,Informative': 'study',

    'Streaming Services': 'entertainment',
    'Games': 'entertainment',
    'Food': 'entertainment',
    'Comedy,Entertainment': 'entertainment',
    'Entertainment,Comedy': 'entertainment',
    'Automobile,Comedy': 'entertainment',
    'Blog,Comedy': 'entertainment',
    'Comedy,Informative': 'entertainment',
    'Tech,Comedy': 'entertainment',
    'Comedy': 'entertainment',
    'Blog,Entertainment': 'entertainment',
    'Entertainment,Blog': 'entertainment',
    'Food,Entertainment': 'entertainment',
    'VideoGames': 'entertainment',
    'Entertainment': 'entertainment',

    'Travel': 'entertainment',
    'Social Networking and Messaging': 'others',
    'News': 'others',
    'Sports': 'entertainment',
    'Photography': 'others',
    'Law and Government': 'others',
    'Health and Fitness': 'others',
    'E-Commerce': 'others',
    'Forums': 'others',
    'Automobile': 'others',
    'Tech,News': 'study',
    'Business/Corporate': 'others',

}

# Apply the mapping to the 'Category' column and filter out 'Adult'
df_combined_reclassified = df_combined[df_combined['Category'] != 'Adult'].copy()
df_combined_reclassified['Broad_Category'] = df_combined_reclassified['Category'].map(category_mapping)

# Drop rows where the mapping resulted in NaN (categories not in the mapping)
df_combined_reclassified.dropna(subset=['Broad_Category'], inplace=True)


# Display the value counts of the new categories
print(df_combined_reclassified['Broad_Category'].value_counts())

Broad_Category
entertainment    1626
others           1094
study            1027
Name: count, dtype: int64


# Task
Classify the 'Category' column into 'study', 'entertainment', and 'others' based on the 'preprocessed_text' column, remove rows classified as 'adult', and then train and evaluate a classification model using the reclassified data.

## Vectorize reclassified data

### Subtask:
Use `TfidfVectorizer` to convert the 'preprocessed_text_str' column of the reclassified dataframe into numerical features.


**Reasoning**:
Transform the preprocessed text data in the reclassified dataframe using TfidfVectorizer and print the shape of the resulting matrix.



In [52]:
tfidf_vectorizer_reclassified = TfidfVectorizer()
tfidf_matrix_reclassified = tfidf_vectorizer_reclassified.fit_transform(df_combined_reclassified['preprocessed_text_str'])

print("TF-IDF matrix shape of reclassified data:", tfidf_matrix_reclassified.shape)

TF-IDF matrix shape of reclassified data: (3747, 46861)


## Split reclassified data

### Subtask:
Split the vectorized data and the new broad categories into training and testing sets.


**Reasoning**:
Split the vectorized reclassified data and the new broad categories into training and testing sets using train_test_split and print the shapes.



In [53]:
X_reclassified = tfidf_matrix_reclassified
y_reclassified = df_combined_reclassified['Broad_Category']

X_train_reclassified, X_test_reclassified, y_train_reclassified, y_test_reclassified = train_test_split(X_reclassified, y_reclassified, test_size=0.2, random_state=42)

print("Shape of X_train_reclassified:", X_train_reclassified.shape)
print("Shape of X_test_reclassified:", X_test_reclassified.shape)
print("Shape of y_train_reclassified:", y_train_reclassified.shape)
print("Shape of y_test_reclassified:", y_test_reclassified.shape)

Shape of X_train_reclassified: (2997, 46861)
Shape of X_test_reclassified: (750, 46861)
Shape of y_train_reclassified: (2997,)
Shape of y_test_reclassified: (750,)


## Train and evaluate model

### Subtask:
Train a classification model (e.g., Logistic Regression, Support Vector Machine, or RandomForestClassifier) on the training data and evaluate its performance using a classification report and accuracy score.


**Reasoning**:
Train a Logistic Regression model and evaluate its performance using a classification report and accuracy score.



In [54]:
model_reclassified = LogisticRegression()
model_reclassified.fit(X_train_reclassified, y_train_reclassified)
predictions_reclassified = model_reclassified.predict(X_test_reclassified)

print("Classification Report (Reclassified Categories):")
print(classification_report(y_test_reclassified, predictions_reclassified))

print("Accuracy Score (Reclassified Categories):")
print(accuracy_score(y_test_reclassified, predictions_reclassified))

Classification Report (Reclassified Categories):
               precision    recall  f1-score   support

entertainment       0.85      0.90      0.88       347
       others       0.91      0.83      0.87       221
        study       0.80      0.80      0.80       182

     accuracy                           0.86       750
    macro avg       0.85      0.84      0.85       750
 weighted avg       0.86      0.86      0.86       750

Accuracy Score (Reclassified Categories):
0.856


## Predict category for new data

### Subtask:
Preprocess the new text data, vectorize it using the fitted TfidfVectorizer, and predict its category using the trained model.

**Reasoning**:
Preprocessing and vectorizing the new text data and using the trained model to predict its category.

In [55]:
# Create a function to predict the category of a new string input
def predict_category(text_input, model, vectorizer, preprocessing_func):
    """
    Predicts the category of a given text string.

    Args:
        text_input (str): The input text string.
        model: The trained classification model.
        vectorizer: The fitted TF-IDF vectorizer.
        preprocessing_func (function): The preprocessing function to apply.

    Returns:
        str: The predicted category.
    """
    # Preprocess the input text
    preprocessed_text = preprocessing_func(text_input)

    # Convert the preprocessed list to a string
    preprocessed_text_str = ' '.join(preprocessed_text)

    # Vectorize the new text using the same fitted TF-IDF vectorizer
    text_tfidf = vectorizer.transform([preprocessed_text_str])

    # Predict the category
    predicted_category = model.predict(text_tfidf)

    return predicted_category[0]

# Create a wrapper function that takes only a string input
def predict_category_from_string(text_input_string):
  """
  Predicts the category of a given text string using the pre-trained model and vectorizer.

  Args:
      text_input_string (str): The input text string.

  Returns:
      str: The predicted category.
  """
  return predict_category(text_input_string, model_reclassified, tfidf_vectorizer_reclassified, preprocessing)




In [56]:
import pickle

# Save the trained model
with open('model_reclassified.pkl', 'wb') as f:
    pickle.dump(model_reclassified, f)

# Save the fitted TF-IDF vectorizer
with open('tfidf_vectorizer_reclassified.pkl', 'wb') as f:
    pickle.dump(tfidf_vectorizer_reclassified, f)

print("Trained model and TF-IDF vectorizer saved as pickle files.")

Trained model and TF-IDF vectorizer saved as pickle files.


In [57]:

predicted_category_example = predict_category_from_string("example")

print("Predicted category for the example text:", predicted_category_example)

Predicted category for the example text: study


In [58]:
import pickle

# Save the predict_category_from_string function
with open('predict_category_function.pkl', 'wb') as f:
    pickle.dump(predict_category_from_string, f)

print("predict_category_from_string function saved as a pickle file.")

predict_category_from_string function saved as a pickle file.


## Load and use the pickled prediction function

### Subtask:
Load the pickled `predict_category_function` and use it to get a prediction for a sample string input.

**Reasoning**:
Demonstrating how to load the pickled function and use it for prediction.

## Summary:

### Data Analysis Key Findings

*   The reclassified data, after vectorization using `TfidfVectorizer`, resulted in a matrix shape of (3747, 46861), indicating 3747 documents and 46861 unique terms.
*   The reclassified data was successfully split into training and testing sets with an 80/20 ratio. The training set has 2997 samples and the testing set has 750 samples.
*   A Logistic Regression model was trained on the reclassified data and achieved an accuracy score of 0.856 on the test set.

### Insights or Next Steps

*   The model shows good performance in classifying the broad categories. Further analysis of the classification report could identify areas for improvement for specific categories.
*   Consider exploring other classification algorithms (e.g., RandomForestClassifier, Support Vector Machines) to potentially improve accuracy or efficiency.
